plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java

```diff
@@ -0,0 +1,309 @@
+            RemoteQueryModifier remoteQueryModifier)
+    {
+        return new JdbcPageSink(session, jdbcOutputTableHandle, jdbcClient, pageSinkId, remoteQueryModifier);
+    }
+
+    protected ConnectorPageSink createUpdateSink(
+            ConnectorSession session,
+            JdbcOutputTableHandle jdbcOutputTableHandle,
+            JdbcClient jdbcClient,
+            ConnectorPageSinkId pageSinkId,
+            RemoteQueryModifier remoteQueryModifier)
+    {
+        ImmutableList.Builder<String> updateColumnNames = ImmutableList.builder();
+        ImmutableList.Builder<Type> updateColumnTypes = ImmutableList.builder();
+        for (int channel : reservedChannelsForUpdate) {
+            updateColumnNames.add(jdbcOutputTableHandle.getColumnNames().get(channel));
+            updateColumnTypes.add(jdbcOutputTableHandle.getColumnTypes().get(channel));
+        }
+        updateColumnNames.addAll(mergeRowIdFieldNames);
+        updateColumnTypes.addAll(mergeRowIdFieldTypes);
+
+        JdbcOutputTableHandle updateOutputTableHandle = new JdbcOutputTableHandle(
+                catalogName,
+                schemaName,
+                tableName,
+                updateColumnNames.build(),
+                updateColumnTypes.build(),
+                Optional.empty(),
+                Optional.empty(),
+                Optional.empty());
+        return new UpdateSink(session, updateOutputTableHandle, jdbcClient, pageSinkId, remoteQueryModifier);
+    }
+
+    protected ConnectorPageSink createDeleteSink(
+            ConnectorSession session,
+            Optional<JdbcOutputTableHandle> jdbcDeleteOutputTableHandle,
+            JdbcClient jdbcClient,
+            ConnectorPageSinkId pageSinkId,
+            RemoteQueryModifier remoteQueryModifier)
+    {
+        if (jdbcDeleteOutputTableHandle.isPresent()) {
+            return new JdbcPageSink(session, jdbcDeleteOutputTableHandle.get(), jdbcClient, pageSinkId, remoteQueryModifier);
+        }
+
+        JdbcOutputTableHandle deleteOutputTableHandle = new JdbcOutputTableHandle(
+                catalogName,
+                schemaName,
+                tableName,
+                mergeRowIdFieldNames,
+                mergeRowIdFieldTypes,
+                Optional.empty(),
+                Optional.empty(),
+                Optional.empty());
+        return new DeleteSink(session, deleteOutputTableHandle, jdbcClient, pageSinkId, remoteQueryModifier);
+    }
+
+    private class UpdateSink
+            extends JdbcPageSink
+    {
+        public UpdateSink(ConnectorSession session, JdbcOutputTableHandle handle, JdbcClient jdbcClient, ConnectorPageSinkId pageSinkId, RemoteQueryModifier remoteQueryModifier)
+        {
+            super(session, handle, jdbcClient, pageSinkId, remoteQueryModifier);
+        }
+
+        @Override
+        protected String getSinkSql(JdbcClient jdbcClient, JdbcOutputTableHandle outputTableHandle, List<WriteFunction> columnWriters)
```

Comment:
I think we should somehow reuse `io.trino.plugin.jdbc.DefaultQueryBuilder#prepareUpdateQuery`